src/sentry/hybridcloud/models/outbox.py

```diff
@@ -219,11 +220,144 @@ def process_shard(self, latest_shard_row: OutboxBase | None) -> Generator[Outbox
+        """
+        using = router.db_for_write(type(self))
+        tags: dict[str, int | str] = {"category": "None", "synchronous": int(is_synchronous_flush)}
+
+        now = timezone.now()
+
+        filters: dict[str, Any] = {}
+        if latest_shard_row is not None:
+            # If we have a latest row, only operate on messages that were created earlier
+            filters = {"id__lte": latest_shard_row.id}
+
+        try:
+            with transaction.atomic(using=using, savepoint=False):
+                # Lock all the rows we're going to process and attempt to reserve them.
+                reserved_ids = list(
+                    self.select_coalesced_messages()
+                    .filter(**filters)
+                    .select_for_update(nowait=True)
+                    .order_by("id")
+                    .values_list("id", flat=True)
+                )
+                if not reserved_ids:
+                    yield None
+                    return
+
+                coalesced: OutboxBase = self.objects.get(id=reserved_ids[-1])
+                first_coalesced: OutboxBase = (
+                    self.objects.filter(id=reserved_ids[0]).first() or coalesced
+                )
+
+                # Reserve all messages in the coalesce group for a short time so they cannot
+                # be selected by concurrent calls.
+                self.objects.filter(
+                    id__in=reserved_ids,
+                ).update(scheduled_for=now + datetime.timedelta(minutes=5))
+        except OperationalError as e:
+            # If concurrent locking is happening on the table, gracefully pass and allow
+            # that work to process.
+            if "LockNotAvailable" in str(e):
+                yield None
+                return
+            else:
+                raise
+
+        tags["category"] = OutboxCategory(self.category).name
+        assert first_coalesced, "first_coalesced incorrectly set for non-empty coalesce group"
+        metrics.timing(
+            "outbox.coalesced_net_queue_time",
+            now.timestamp() - first_coalesced.date_added.timestamp(),
+            tags=tags,
+        )
+
+        # Yield the latest message for processing outside of a transaction
+        yield coalesced
+
+        # If the block didn't raise, delivery was successful, and we
+        # can delete messages to acknowledge them.
+        deleted_count = 0
+        with transaction.atomic(using=using):
+            # The last message in a coalesce group should be preserved if we're
+            # skipping the shard, as we need to delay signals, and keep
+            # backlogs shorter, but not lose updates.
+            cleanup_ids = reserved_ids[:-1]
+            if cleanup_ids:
+                self.objects.filter(id__in=cleanup_ids).delete()
+                deleted_count = len(cleanup_ids)
+            if not self.should_skip_shard():
```

Comment:
If `self.should_skip_shard()` is true, then we never delete the last outbox. We always push its `scheduled_for` out by 5 minutes.